File: cli/src/contract_status/code.rs
/*
 * Copyright 2018 Fluence Labs Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use contract_func::ContractCaller;
use std::error::Error;
use web3::types::{Address, H256};

#[derive(Serialize, Deserialize, Debug, Getters)]
pub struct Code {
    storage_hash: H256,
    storage_receipt: H256,
    cluster_size: u8,
    developer: Address,
}

impl Code {
    pub fn new(
        storage_hash: H256,
        storage_receipt: H256,
        cluster_size: u8,
        developer: Address,
    ) -> Code {
        Code {
            storage_hash,
            storage_receipt,
            cluster_size,
            developer,
        }
    }
}

pub fn get_enqueued_codes(contract: &ContractCaller) -> Result<Vec<Code>, Box<Error>> {
    let (storage_hashes, storage_receipts, cluster_sizes, developers): (
        Vec<H256>,
        Vec<H256>,
        Vec<u64>,
        Vec<Address>,
    ) = contract.query_contract("getEnqueuedCodes", ())?;

    let mut codes: Vec<Code> = Vec::new();
    for i in 0..storage_hashes.len() {
        let code = Code::new(
            storage_hashes[i],
            storage_receipts[i],
            cluster_sizes[i] as u8,
            developers[i],
        );
        codes.push(code);
    }

    Ok(codes)
}

File: monitoring/src/code.ts
/*
 * Copyright (C) 2017  Fluence Labs Limited
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {Network} from "../types/web3-contracts/Network";

/**
 * A code is a WASM file. It can be deployed on a real-time cluster and run.
 * In Fluence contract it represents as a Swarm address to the WASM file
 * and a requirement of how many nodes will be in the cluster.
 */
export interface Code {
    code_address: string,
    storage_receipt: string,
    cluster_size: number,
    developer: string
}

/**
 * Gets list of enqueued codes from Fluence contract
 */
export async function getEnqueuedCodes(contract: Network): Promise<Code[]> {
    let unparsedCodes = await contract.methods.getEnqueuedCodes().call();

    let codeAddresses = unparsedCodes["0"];
    let storageReceipts = unparsedCodes["1"];
    let clusterSizes = unparsedCodes["2"];
    let developers = unparsedCodes["3"];

    return parseCodes(codeAddresses, storageReceipts, clusterSizes, developers);
}

/**
 * Collects codes from response format of Fluence contract.
 */
export function parseCodes(codeAddresses: string[],
                    storageReceipts: string[],
                    clusterSizes: string[],
                    developers: string[]): Code[] {
    let codes: Code[] = [];

    codeAddresses.forEach((address, index) => {
        let code: Code = {
            code_address: address,
            storage_receipt: storageReceipts[index],
            cluster_size: parseInt(clusterSizes[index]),
            developer: developers[index]
        };
        codes.push(code);
    });
    return codes;
}
